sw/host/opentitanlib/src/transport/ti50emulator/i2c.rs

```diff
@@ -2,14 +2,338 @@
+        if buf.len() > 2 {
+            buf[0] = TI50_I2C_BUS_STOP;
+            buf[1] = TI50_I2C_MAGIC;
+            return Ok(2);
+        }
+        bail!("BusControlPacket invalid buffer length");
+    }
+
+    /// Write byte representation of start packet
+    pub fn emit_data(data: &[u8], buf: &mut [u8]) -> Result<usize> {
+        if buf.len() > 2 {
+            buf[0] = TI50_I2C_BUS_DATA;
+            buf[1] = data.len() as u8;
+            return Ok(2);
+        }
+        bail!("BusControlPacket invalid buffer length");
+    }
+}
+
+/// Structure representing Emulated I2C BUS
+pub struct Ti50I2cBus {
+    inner: Rc<RefCell<Inner>>,
+    // This socket is valid as long as SubProcess is running.
+    socket: RefCell<Option<UnixStream>>,
+    // full path to socket file
+    path: PathBuf,
+    // last SubProcess ID
+    last_id: Cell<u64>,
+}
+
+impl Ti50I2cBus {
+    /// Create new instance of [`Ti50I2cBus`] based on provided parameters.
+    pub fn open(ti50: &Ti50Emulator, path: &str) -> Result<Self> {
+        let soc_path = ti50.inner.borrow().process.get_runtime_dir().join(path);
+        Ok(Self {
+            inner: Rc::clone(&ti50.inner),
+            socket: RefCell::default(),
+            path: soc_path,
+            last_id: Cell::new(EMULATOR_INVALID_ID),
+        })
+    }
+
+    /// Function re-connect socket to `SubProcess` when detect
+    /// that process was restarted.
+    pub fn reconnect(&self) -> Result<()> {
+        let mut socket = self.socket.borrow_mut();
+        let id = self.inner.borrow_mut().process.get_id();
+        if self.last_id.get() != id {
+            *socket = Some(UnixStream::connect(&self.path)?);
+            self.last_id.set(id);
+        }
+        Ok(())
+    }
+
+    /// Function check if I2C transaction should be performed on DUT.
+    pub fn check_state(&self) -> Result<()> {
+        let process = &mut self.inner.borrow_mut().process;
+        process.update_status()?;
+        match process.get_state() {
+            EmuState::On => Ok(()),
+            state => Err(I2cError::Generic(format!(
+                "Operation not supported in Emulator state: {}",
+                state
+            ))
+            .into()),
+        }
+    }
+
+    /// Function check if emulated I2C BUS is busy if so return error.
+    pub fn is_bus_busy(fd: &mut UnixStream) -> Result<()> {
+        let mut peek_buf: [u8; 8] = [0; 8];
+        match fd.read(&mut peek_buf) {
+            Ok(0) => {
+                return Ok(());
+            }
+            Ok(_n) => {
+                bail!(I2cError::Busy);
+            }
+            Err(err) => {
+                if err.kind() != ErrorKind::WouldBlock {
+                    return Err(err.into());
+                }
+                return Ok(());
+            }
+        }
+    }
+
+    /// Function try to receive data from unix socket.
+    pub fn recive(fd: &mut UnixStream, data: &mut [u8], deadline: Instant) -> Result<usize> {
```

Comment:
There is something fundamentally missing here.

If the simulated Dauntless chip has 63 bytes ready to be read by the host.  But the host chooses to read only 16 bytes and end the transaction.  In the code on the chip side, the I2CHwSlaveClient callback function expectes to be passed 16 as a parameter, indicating how many bytes were actually read by the host.

As far as I can tell, this implementation does not make any attempt at propagating that information to the emulated chip.

This method needs to start out by transmitting a few bytes to the emulation, probably a new TI50_I2C_BUS_RECV_DATA followed by 16-bit length.  Once the emulation knows how many bytes the host wants to receive, it will transmit that exact number of bytes through the pipe, and provide the number to the callback.

Reply:
I rewrote the whole thing to take this fact into account. Due to the fact that we do not have to check bus state, the code has been greatly simplified.